libraries/tickfs/src/flash_controller.rs

```diff
@@ -0,0 +1,81 @@
+//! The Flash Controller interface with hardware
+
+use crate::error_codes::ErrorCode;
+
+/// Implementation required for the flash controller hardware. This
+/// should read, write and erase flash from the hardware using the
+/// flash controller.
+///
+/// This is the public trait for the Flash controller implementation.
+///
+/// The size of the regions (pages) must be the smallest size that can be
+/// erased in a single operation. This must match the `region_size`
+/// and the length of the `read_buffer` slice used in TickFS.
+///
+/// The start and end address of the FlashController must be aligned
+/// to the size of regions.
+/// All `region_number`s and `address`es are offset from zero. If you
+/// want to use flash that doesn't start at zero, or is a partition
+/// offset from the start of flash you will need to add that offset
+/// to the values in your implementation.
+///
+/// The boiler plate for an implementation will look something like this
+///
+/// ```rust
+/// use tickfs::error_codes::ErrorCode;
+/// use tickfs::flash_controller::FlashController;
+///
+/// #[derive(Default)]
+/// struct FlashCtrl {}
+///
+/// impl FlashCtrl {
+///     fn new() -> Self {
+///         Self { /* fields */ }
+///     }
+/// }
+///
+/// impl FlashController for FlashCtrl {
+///     fn read_region(&self, region_number: usize, offset: usize, buf: &mut [u8]) -> Result<(), ErrorCode> {
+///         unimplemented!()
+///     }
+///
+///     fn write_region(&self, address: usize, buf: &[u8]) -> Result<(), ErrorCode> {
+///         unimplemented!()
+///     }
+///
+///     fn erase_region(&self, region_number: usize) -> Result<(), ErrorCode> {
+///         unimplemented!()
+///     }
+/// }
+/// ```
+pub trait FlashController {
+    /// This function must read the data from the flash region specified by
+    /// `region_number` into `buf`. The length of the data read should be the
+    /// same length as buf. `offset` indicates an offset into the region that
+    /// should be read.
+    ///
+    /// On success it should return nothing, on failure it
+    /// should return ErrorCode::ReadFail.
+    fn read_region(
+        &self,
+        region_number: usize,
+        offset: usize,
+        buf: &mut [u8],
+    ) -> Result<(), ErrorCode>;
```

Comment:
Is `buf` required to end up entirely within a single region?

Reply:
So right now buf will contain the entire region size. In future it might be worth allowing a smaller buf and read part way through a region.